### PR TITLE
Add task execution cleanup --all option

### LINKED ISSUE #3902: Add support to cleanup task executions
As a user, I'd like the ability to clean up all the task executions, so when I'm in an active dev mode on CF, I have the ability to clean slate the task footprint quickly.

Currently, we have the support to clean up the execution by a particular task execution ID.
> dataflow:>task execution cleanup --id

It would be useful to expose the following option:
> dataflow:>task execution cleanup --all

**Acceptance:**
- Issuing `--all` would clean up all the task executions that aren't actively running or in use by some other state in SCDF. 
- Attempt to clean up any and all and not terminate if any of the executions are not in an inconsistent state or if it is not deletable.
- Show the resulting cleanup count in the response.
- If something fails, show the stacktrace as part of the resulting response.

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- This `all` option would remove all the task execution resources along with the task execution records
 - Allow option to remove only the `completed` task executions
 - Add controller endpoint to retrieve total task executions for performance reasons
 - Add new interface methods into TaskExecutionService to retrieve the task executions using the filter option
 - Add/update tests

Resolves #3902

### CODE REVIEW

**Comment by ilayaperumalg** [comment]
The UI at the CLI would like below and looking forward to get your opinions as well:

```
dataflow:>task execution cleanup --all --completed-executions
About to delete 3 completed task executions and related records. This operation can not be reverted. Are you sure (y/n)?  [y, n]: n
Cleanup process is canceled
dataflow:>task execution cleanup --all
About to delete 4 task executions and related records. This operation can not be reverted. Are you sure (y/n)?  [y, n]: y
Request to clean up resources for task executions has been submitted
dataflow:>task execution list
╔═════════╤══╤══════════╤════════╤═════════╗
║Task Name│ID│Start Time│End Time│Exit Code║
╚═════════╧══╧══════════╧════════╧═════════╝

```

**Comment by jvalkeal** [comment]
Think it fails with ctr:
```
dataflow:>task list
╔═════════╤══════════════════════════════╤═══════════╤═══════════╗
║Task Name│       Task Definition        │description│Task Status║
╠═════════╪══════════════════════════════╪═══════════╪═══════════╣
║ctr1     │t1: timestamp && t2: timestamp│           │COMPLETE   ║
║ctr1-t1  │t1: timestamp                 │           │COMPLETE   ║
║ctr1-t2  │t2: timestamp                 │           │COMPLETE   ║
║t1       │timestamp                     │           │COMPLETE   ║
║t2       │timestamp                     │           │COMPLETE   ║
╚═════════╧══════════════════════════════╧═══════════╧═══════════╝

dataflow:>task execution cleanup --all --completed-executions
About to delete 4492 completed task executions and related records. This operation can not be reverted. Are you sure (y/n)?  [y, n]: y
Command failed org.springframework.cloud.dataflow.rest.client.DataFlowClientException: Cannot delete non-parent TaskExecutions with the following ids: 15, 16, 17, 18, 20, .....
Cannot delete non-parent TaskExecutions with the following ids: 15, 16, 17, 18, 20, .....
org.springframework.cloud.dataflow.rest.client.DataFlowClientException: Cannot delete non-parent TaskExecutions with the following ids: 15, 16, 17, 18, 20,.....
    at org.springframework.cloud.dataflow.rest.client.VndErrorResponseErrorHandler.handleError(VndErrorResponseErrorHandler.java:65)
```

server:
```
[nio-9393-exec-8] o.s.c.d.s.c.RestControllerAdvice         : Caught exception while handling a request: Cannot delete non-parent TaskExecutions with the following ids: 15, 16, 17, ......
```

**Comment by ilayaperumalg** [comment]
@jvalkeal Fixed the issue with the composed task apps deletion with this change 4304503

**Comment by jvalkeal** [comment]
We'd probably need to modify DELETE to be more lined what you can do with GET for a same path. User would probably like to have an option to delete executions only for a particular task and thus having just a delete for all is a bit too limited.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskExecutionController.java [hunk 1]
```diff
@@ -215,6 +216,12 @@ public TaskExecutionResource view(@PathVariable("id") long id) {
 		return resources;
 	}
 
+	@RequestMapping(method = RequestMethod.GET, params = "count")
+	@ResponseStatus(HttpStatus.OK)
+	public TaskExecutionsResource getTaskExecutionIds(@RequestParam(defaultValue = "false", name="completed") boolean completed) {
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskExecutionController.java:221]: This also has other issues like:
```
http ":9393/tasks/executions?name=t1&count"

HTTP/1.1 500 
Connection: close
Content-Type: application/hal+json
Date: Tue, 23 Feb 2021 09:08:03 GMT
Transfer-Encoding: chunked

{
    "_embedded": {
        "errors": [
            {
                "_links": {
                    "self": {
                        "href": "/"
                    }
                },
                "logref": "IllegalStateException",
                "message": "Ambiguous handler methods mapped for '/tasks/executions': {public org.springframework.hateoas.PagedModel org.springframework.cloud.dataflow.server.controller.TaskExecutionController.retrieveTasksByName(java.lang.String,org.springframework.data.domain.Pageable,org.springframework.data.web.PagedResourcesAssembler), public org.springframework.cloud.dataflow.rest.resource.TaskExecutionsResource org.springframework.cloud.dataflow.server.controller.TaskExecutionController.getTaskExecutionIds(boolean)}"
            }
        ]
    }
}

```

Comment (jvalkeal) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskExecutionController.java:221]: Also these both gives same response:
```
http ":9393/tasks/executions?count&completed=true"
http ":9393/tasks/executions?count&completed=false"
```

Comment (jvalkeal) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskExecutionController.java:221]: Ah it's inclusion, so gives different count when things are running.

Comment (ilayaperumalg) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskExecutionController.java:221]: Yeah, I see the ambiguity and will fix it.

**Comment by ilayaperumalg** [comment]
Pushed the changes to address:

- task execution cleanup by task name
- task execution controller endpoint to have `/info` as an unique REST endpoint to retrieve executions info

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskExecutionController.java [hunk 2]
```diff
@@ -155,6 +156,16 @@ public TaskExecutionController(TaskExplorer explorer, TaskExecutionService taskE
 		return assembler.toModel(result, this.taskAssembler);
 	}
 
+	@RequestMapping(value= "", method = RequestMethod.GET, params = "info")
+	@ResponseStatus(HttpStatus.OK)
+	public TasksInfoResource getTaskExecutionsInfo(@RequestParam(defaultValue = "true") Boolean info,
+												   @RequestParam(defaultValue = "false", name="completed") Boolean completed,
+												   @RequestParam(defaultValue = "", name="name") String taskName) {
+		return (info == true) ? new TaskExecutionsAssembler().toModel(
+				this.taskExecutionService.getAllTaskExecutionsCount(completed, taskName)) : new TasksInfoResource();
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TaskExecutionController.java:165]: Theoretical NPE as `completed` is `Boolean` and `getAllTaskExecutionsCount` takes `boolean`.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/TaskExecutionService.java [hunk 3]
```diff
@@ -75,4 +75,20 @@
...
+	 * @param onlyCompleted filter by completed task executions
+	 * @param taskName the task name, if null then retrieve all the tasks
+	 * @since 2.8
+	 */
+	Set<Long> getAllTaskExecutionIds(boolean onlyCompleted, String taskName);
+
+	/**
+	 * Returns the count of all the task execution IDs with the option to include only the completed task executions.
+	 * @param onlyCompleted filter by completed task executions
+	 *@param taskName the task name, if null then retrieve all the tasks
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/TaskExecutionService.java:90]: Looks like we don't have checkstyle for missing space.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowTaskExecutionDao.java [hunk 4]
```diff
@@ -165,4 +194,68 @@ private String getQuery(String base) {
...
+		if (StringUtils.hasText(taskName)) {
+			queryParameters = new MapSqlParameterSource()
+					.addValue("taskName", taskName, Types.VARCHAR);
+			QUERY = (onlyCompleted) ? GET_COMPLETED_TASK_EXECUTIONS_COUNT_BY_TASK_NAME : GET_ALL_TASK_EXECUTIONS_COUNT_BY_TASK_NAME;
+		}
+		else {
+			QUERY = (onlyCompleted) ? GET_COMPLETED_TASK_EXECUTIONS_COUNT: GET_ALL_TASK_EXECUTIONS_COUNT;
+		}
+		try {
+			//return this.simpleJdbcTemplate.query(getQuery(QUERY), queryParameters, Integer.class);
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/repository/JdbcDataflowTaskExecutionDao.java:211]: Can remove.

File: spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/TaskTemplate.java [hunk 5]
```diff
@@ -246,6 +253,26 @@ public void cleanup(long id, boolean removeData) {
...
 	}
 
+	@Override
+	public void cleanupAllTaskExecutions(boolean completed, String taskName) {
+		String uriTemplate = this.executionsLink.getHref() + "?action=CLEANUP,REMOVE_DATA";
+		if (completed) {
+			uriTemplate = uriTemplate + "&completed=true";
+		}
+		if (StringUtils.hasText(taskName)) {
+			uriTemplate = uriTemplate + "&name="+taskName;
```
Comment (jvalkeal) [inline spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/TaskTemplate.java:263]: Looks like we don't have checkstyle for missing space.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TasksInfoController.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+	 * @param taskExecutionService used to launch tasks
+	 */
+	public TasksInfoController(TaskExecutionService taskExecutionService) {
+		Assert.notNull(taskExecutionService, "taskExecutionService must not be null");
+		this.taskExecutionService = taskExecutionService;
+	}
+
+	@RequestMapping(value= "executions", method = RequestMethod.GET)
+	@ResponseStatus(HttpStatus.OK)
+	public TasksInfoResource getInfo(@RequestParam(defaultValue = "false", name="completed") Boolean completed,
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/controller/TasksInfoController.java:65]: Possible NPE with `completed`

**Pull request merged**